MobileSdk/src/main/java/com/spruceid/mobile/sdk/GattClient.kt

```diff
@@ -333,101 +348,300 @@ class GattClient(private var callback: GattClientCallback,
+                    return
                 }
+                payload.write(buf, 0, buf.count())
+
+                dprint("Currently have ${buf.count()} bytes.")
+
+                requestTimestamp = TimeSource.Monotonic.markNow()
+
+                Executors.newSingleThreadScheduledExecutor()
+                    .schedule({
+                                  val curtime = TimeSource.Monotonic.markNow()
+                                  if ((curtime - requestTimestamp) > 500.milliseconds) {
+                                      val message = payload.toByteArray()
+
+                                      dprint("Request complete: ${message.count()} bytes.")
+                                      callback.onMessageReceived(message)
+                                  }
+                              }, 500, TimeUnit.MILLISECONDS)
+
+            } catch (e: IOException) {
+                reportError("Error on listening input stream from socket L2CAP: ${e}")
                 return
             }
+        }
+    }
+
+    /**
+     * Thread for writing the response via L2CAP.
+     */
+    fun writeResponse() {
+        val outStream = l2capSocket!!.outputStream
+        try {
+            while (true) {
+                var message: ByteArray?
+                try {
+                    message = responseData.poll(500, TimeUnit.MILLISECONDS)
+                    if (message == null) {
+                        continue
+                    }
+                    if (message.size == 0) {
+                        break
+                    }
+                } catch (e: InterruptedException) {
+                    continue
+                }
+                outStream.write(message)
+            }
+        } catch (e: IOException) {
+            reportError("Error writing response via L2CAP socket: ${e}")
+        }
+
+        try {
+            // TODO: This is to work around a bug in L2CAP
+            Thread.sleep(1000)
+            l2capSocket!!.close()
+        } catch (e: IOException) {
+            reportError("Error closing socket: ${e}")
+        } catch (e: InterruptedException) {
+            reportError("Error closing socket: ${e}")
+        }
+    }
+
+    /**
+     * Set notifications for a characteristic.  This process is rather more complex than you'd think it would
+     * be, and isn't complete until onDescriptorWrite() is hit; it triggers an async action.
+     */
+    private fun enableNotification(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic?, name: String) {
+        dprint("Enabling notifications on ${name}")
 
-            if (!gatt.setCharacteristicNotification(characteristicServer2Client, true)) {
-                callback.onError(Error("Error setting notification on Server2Client."))
+        if (characteristic == null) {
+            reportError("Error setting notification on ${name}; is null.")
+            return
+        }
+
+        dprint("-- ind: ${characteristic.getProperties()}")
```

Comment:
```suggestion
        reportLog("-- ind: ${characteristic.getProperties()}")
```

Reply:
Removed; this was some debug scaffolding.